Igor Yu. Zhbanov wrote:
> I am using rsync compiled with Cygwin on windows.
> I must call rsync from the *.bat script (I don't want to use a bash on Windows)
> and I have noticed that in the case when program compiled by Cygwin crashes
> via segmentation fault and default SIGSEGV handler is called, then it
> terminates process with exit status 0 as I see it from my *.bat script.
> (But if I invoke a program from bash (compiled with Cygwin too) I will see
> error code 139 as expected.)
>
> It is a Cygwin's problem, not an rsync's, but to use it on windows and
> to distinguish situations when rsync crashes and when it exits normally,
> I have written signal handler which terminates process with code 50.
> You may use conventional 139. Also signal handler writes corresponding
> message to log file.
>
> By the way. When I terminate rsync in daemon mode by pressing Control-C,
> it writes an error to log. May be this is not an error but info or notice?

I'm not sure I like this, but if you run into the cygwin problem, this might
prove helpful.

To use this patch, run these commands for a successful build:

    patch -p1 <patches/catch_crash_signals.diff
    ./configure                                  (optional if already run)
    make

based-on: 0ac7ebceef70417355f25daf9e2fd94e84c49749
diff --git a/errcode.h b/errcode.h
--- a/errcode.h
+++ b/errcode.h
@@ -47,6 +47,8 @@
 #define RERR_TIMEOUT    30      /* timeout in data send/receive */
 #define RERR_CONTIMEOUT 35      /* timeout waiting for daemon connection */
 
+#define RERR_WECRASHED	 50      /* We have crashed. */
+
 /* Although it doesn't seem to be specified anywhere,
  * ssh and the shell seem to return these values:
  *
diff --git a/log.c b/log.c
--- a/log.c
+++ b/log.c
@@ -92,6 +92,7 @@ struct {
 	{ RERR_TERMINATED , "sibling process terminated abnormally" },
 	{ RERR_SIGNAL1    , "received SIGUSR1" },
 	{ RERR_SIGNAL     , "received SIGINT, SIGTERM, or SIGHUP" },
+	{ RERR_WECRASHED  , "rsync caught a CRASH-causing signal" },
 	{ RERR_WAITCHILD  , "waitpid() failed" },
 	{ RERR_MALLOC     , "error allocating core memory buffers" },
 	{ RERR_PARTIAL    , "some files/attrs were not transferred (see previous errors)" },
diff --git a/main.c b/main.c
--- a/main.c
+++ b/main.c
@@ -216,8 +216,11 @@ static void wait_process_with_flush(pid_t pid, int *exit_code_ptr)
 			*exit_code_ptr = RERR_TERMINATED;
 		else
 			*exit_code_ptr = RERR_WAITCHILD;
-	} else
+	} else {
 		*exit_code_ptr = WEXITSTATUS(status);
+		if (*exit_code_ptr == RERR_WECRASHED)
+			*exit_code_ptr = RERR_CRASHED;
+	}
 }
 
 void write_del_stats(int f)
@@ -1631,6 +1634,14 @@ void remember_children(UNUSED(int val))
 				break;
 			}
 		}
+		if (WIFSIGNALED(status)) {
+			rprintf(FLOG,
+				"rsync error: (1) Child proccess has unexpectedly died with signal %d\n",
+				WTERMSIG(status));
+		} else if (WIFEXITED(status) && WEXITSTATUS(status) == RERR_WECRASHED) {
+			rprintf(FLOG,
+				"rsync error: (1) Child proccess has CRASHED.\n");
+		}
 	}
 #endif
 #ifndef HAVE_SIGACTION
@@ -1684,6 +1695,13 @@ static void rsync_panic_handler(UNUSED(int whatsig))
 }
 #endif
 
+static void rsync_crash_handler(UNUSED(int whatsig))
+{
+	log_exit(RERR_WECRASHED, __FILE__, __LINE__);
+	logfile_close();
+	_exit(RERR_WECRASHED);
+}
+
 static void unset_env_var(const char *var)
 {
 #ifdef HAVE_UNSETENV
@@ -1700,7 +1718,6 @@ static void unset_env_var(const char *var)
 #endif
 }
 
-
 int main(int argc,char *argv[])
 {
 	int ret;
@@ -1724,6 +1741,11 @@ int main(int argc,char *argv[])
 	SIGACTMASK(SIGFPE, rsync_panic_handler);
 	SIGACTMASK(SIGABRT, rsync_panic_handler);
 	SIGACTMASK(SIGBUS, rsync_panic_handler);
+#else
+	SIGACTMASK(SIGSEGV, rsync_crash_handler);
+	SIGACTMASK(SIGFPE, rsync_crash_handler);
+	SIGACTMASK(SIGABRT, rsync_crash_handler);
+	SIGACTMASK(SIGBUS, rsync_crash_handler);
 #endif
 #ifdef SIGINFO
 	SIGACTMASK(SIGINFO, siginfo_handler);
diff --git a/socket.c b/socket.c
--- a/socket.c
+++ b/socket.c
@@ -522,7 +522,17 @@ int is_a_socket(int fd)
 static void sigchld_handler(UNUSED(int val))
 {
 #ifdef WNOHANG
-	while (waitpid(-1, NULL, WNOHANG) > 0) {}
+	int status;
+	while (waitpid(-1, &status, WNOHANG) > 0) {
+		if (WIFSIGNALED(status)) {
+			rprintf(FLOG,
+				"rsync error: (3) Child proccess has unexpectedly died with signal %d\n",
+				WTERMSIG(status));
+		} else if (WIFEXITED(status) && WEXITSTATUS(status) == RERR_WECRASHED) {
+			rprintf(FLOG,
+				"rsync error: (3) Child proccess has CRASHED.\n");
+		}
+	}
 #endif
 #ifndef HAVE_SIGACTION
 	signal(SIGCHLD, sigchld_handler);
